Mathlib/Tactic/Bound.lean

```diff
@@ -0,0 +1,264 @@
+
+-- 1 ≤, ≤ 1
+attribute [bound] Nat.one_le_cast_of_le one_le_mul_of_one_le_of_one_le
+
+-- ≤
+attribute [bound] le_abs_self neg_abs_le neg_le_neg tsub_le_tsub_right mul_le_mul_of_nonneg_left
+  mul_le_mul_of_nonneg_right le_add_of_nonneg_right le_add_of_nonneg_left le_mul_of_one_le_right
+  mul_le_of_le_one_right sub_le_sub add_le_add mul_le_mul
+
+-- <
+attribute [bound] Nat.cast_pos_of_pos neg_lt_neg sub_lt_sub_left sub_lt_sub_right add_lt_add_left
+  add_lt_add_right mul_lt_mul_left_of_pos_of_lt mul_lt_mul_right_of_pos_of_lt
+
+-- min and max
+attribute [bound] min_le_right min_le_left le_max_left le_max_right le_min max_le lt_min max_lt
+
+-- Memorize a few constants to avoid going to `norm_num`
+attribute [bound] zero_le_one zero_lt_one zero_le_two zero_lt_two
+
+/-!
+### Forward rules for `bound`
+-/
+
+-- Bound applies `le_of_lt` to all hypotheses
+attribute [bound_forward] le_of_lt
+
+/-!
+### Guessing rules: when we don't know how to recurse
+-/
+
+section Guessing
+
+variable {α : Type} [LinearOrder α] {a b c : α}
+
+-- `min` and `max` guessing lemmas
+lemma le_max_of_le_left_or_le_right : a ≤ b ∨ a ≤ c → a ≤ max b c := le_max_iff.mpr
+lemma lt_max_of_lt_left_or_lt_right : a < b ∨ a < c → a < max b c := lt_max_iff.mpr
+lemma min_le_of_left_le_or_right_le : a ≤ c ∨ b ≤ c → min a b ≤ c := min_le_iff.mpr
+lemma min_lt_of_left_lt_or_right_lt : a < c ∨ b < c → min a b < c := min_lt_iff.mpr
+
+-- Register guessing rules
+attribute [bound]
+  -- Which side of the `max` should we use as the lower bound?
+  le_max_of_le_left_or_le_right
+  lt_max_of_lt_left_or_lt_right
+  -- Which side of the `min` should we use as the upper bound?
+  min_le_of_left_le_or_right_le
+  min_lt_of_left_lt_or_right_lt
+
+end Guessing
+
+/-!
+### Closing tactics
+
+TODO: Kim Morrison noted that we could check for `ℕ` or `ℤ` and try `omega` as well.
+-/
+
+/-- Close numerical goals with `norm_num` -/
+def boundNormNum : Aesop.RuleTac :=
+  Aesop.SingleRuleTac.toRuleTac fun i => do
+    let tac := do Mathlib.Meta.NormNum.elabNormNum .missing .missing .missing
+    let goals ← Lean.Elab.Tactic.run i.goal tac |>.run'
+    if !goals.isEmpty then failure
+    return (#[], none, some .hundred)
+attribute [aesop unsafe 10% tactic (rule_sets := [Bound])] boundNormNum
+
+/-- Close numerical and other goals with `linarith` -/
+def boundLinarith : Aesop.RuleTac :=
+  Aesop.SingleRuleTac.toRuleTac fun i => do
+    Linarith.linarith false [] {} i.goal
+    return (#[], none, some .hundred)
+attribute [aesop unsafe 5% tactic (rule_sets := [Bound])] boundLinarith
+
+/-!
+### `bound` tactic implementation
+-/
+
+/-- Add each provided hypothesis `x` to the context, as by with `have := x` -/
+def addHyps (xs : Array Syntax) : TacticM Unit :=
+  if xs.isEmpty then pure () else Tactic.withMainContext do
+    for x in xs do
+      let v ← elabTerm x none
+      let t ← inferType v
+      liftMetaTactic fun g ↦ do
+        let g ← g.assert `h t v
+        let (_, g) ← g.intro1P
```

Comment:
Use `Lean.MVarId.note`

Reply:
Obsolete, since with your other suggestion this function vanished.